### DISCUSSION TITLE
Compose ViewPager with children

### DISCUSSION THREAD

**Discussion by DatL4g** [arkivanov/Decompose#283] [Q&A]
Is it possible to create a viewpager in compose (using [Accompanist-Pager](https://google.github.io/accompanist/pager/)) with decompose children?

I got a parent screen for example with a childStack of two views.
So the main screen either displays the first or the second child.

Main screen compose
```kotlin
Column {
    Text(text = "Hello World")

    Children(
        stack = component.childStack,
        animation = stackAnimation(fade())
    ) { child ->
        child.instance.render()
    }
}
```

Now I want to have a horizontal viewpager of these children. 

```kotlin
Column {
    Text(text = "Hello World")

    HorizontalPager(count = 2) { position: Int ->
        // how to display the child in here, depending on the position?
    }
}
```

**Comment by arkivanov** [comment]
For pager you definitely don't need `Child Stack`. Pager is a list, not a stack. For a small amount of fixed child components you can just create them directly using `childContext` function. 

```kotlin
val children: List<...> = 
    listOf(
        ComponentA(childContext(key = "childA")),
        ComponentB(childContext(key = "childB")),
    )

...

HorizontalPager(count = children.size) { index ->
    // Render children[index]
}
```

**Comment by DatL4g** [comment]
Pager is neither a classic stack nor a basic list, it's something in between.

A classic stack creates and renders 1 component at a time.
A basic list creates all components and renders how many you want to.

A pager however only needs a max of 3 created and rendered components.
Let's say your on position 2 in a pager, so you need position 1 and position 3 to be created and rendered as well.

If we have for example 50 items we don't want all those items to be created if we just need 3 of them.
Each item may have it's own background logic so we don't want to run havy background tasks for all items, we only want to run 3 background tasks.

In my example it's totally fine to use a list, however maybe it's worth to think about an implementation where you can specify the amount of created components

**Reply by arkivanov** [reply]
Sure, for more complex navigation there is [Generic Navigation](https://arkivanov.github.io/Decompose/navigation/children/overview/) which allows almost any kind of navigation.
